### PR TITLE
Dihard3 recipe (and other minor changes)

### PR DESCRIPTION
**Pull request by desh2608**
1. **Recipe for DIHARD III** (LDC2020E12, LDC2021E02) (dataset is not yet available on LDC)
Usage:
```python
from lhotse.recipes import prepare_dihard3

# prepare both dev and eval
manifests = prepare_dihard3('/export/data/LDC2020E12', '/export/data/LDC2021E02')

# prepare only dev
manifests = prepare_dihard3('/export/data/LDC2020E12', None)
```

CLI usage:
```
Usage: lhotse prepare dihard3 [OPTIONS] OUTPUT_DIR

  DIHARD3 data preparation.

Options:
  --dev PATH
  --eval PATH
  --uem / --no-uem        Specify whether or not to create UEM supervision
                          [default: True]

  -j, --num-jobs INTEGER  Number of jobs to scan corpus directory for
                          recordings.  [default: 1]

  --help                  Show this message and exit.  [default: False]
```

The manifests include recording-wise metadata in the supervisions such as `in_core`, `lang`, `source`, `domain`. For example, to filter the supervisions and get only from the `restaurant` domain, you can do:
```python
manifests['dev']['supervisions'].filter(lambda s: s.custom['domain'] == 'restaurant')
``` 
2. **UEM handling in DiarizationDataset**  
Usage:
```python
from lhotse import CutSet
from lhotse.dataset import DiarizationDataset

cuts = CutSet.from_manifests(recordings=manifests['dev']['recordings'], supervisions=manifests['dev']['supervisions'])
dataset = DiarizationDataset(cuts, uem=manifests['dev']['uem'])
```

3. **Minor change in AMI recipe: option to provide path for annotations**
Previous usage:
```python
from lhotse.recipes import download_ami, prepare_ami

download_ami('/export/data/AMI/')
manifests = prepare_ami('/export/data/AMI', output_dir='data')
```

New usage:
```python
from lhotse.recipes import download_ami, prepare_ami

download_ami('/export/data/AMI/', annotations_dir='/export/data/AMI_annotations')
manifests = prepare_ami('/export/data/AMI', annotations_dir='/export/data/AMI_annotations', output_dir='data')
```
When no `annotations_dir` is provided, it falls back on the `corpus_dir`, as in the old usage.

### CODE REVIEW

File: lhotse/dataset/diarization.py
Comment (pzelasko) [inline lhotse/dataset/diarization.py:69]: If you expect `uem` to be large, it might makes sense to use an interval tree instead (see e.g. `index_supervisions`). But for small sets it should be quick anyway. You might want to consider `overlaps` or `overspans` from `lhotse.utils` for the comparison operations (the intent is more immediately readable)

File: lhotse/recipes/ami.py
Comment (pzelasko) [inline lhotse/recipes/ami.py:293]: I am not convinced that these auto-formatting changes are good, I find the older version much more readable. If you really want to reformat this file, I think there is an option in black to exclude some code lines

File: lhotse/recipes/dihard3.py
Comment (pzelasko) [inline lhotse/recipes/dihard3.py:44]: I recently added `RecordingSet.from_dir` that can do this in parallel if it makes sense

**Review by pzelasko — COMMENTED** [review]
Looks good to me! I left you some minor comments

File: lhotse/dataset/diarization.py
Comment (desh2608) [reply lhotse/dataset/diarization.py:69]: Cool! I wanted to use interval tree too but then I got lazy --- `index_supervisions` should be good.

File: lhotse/recipes/ami.py
Comment (desh2608) [reply lhotse/recipes/ami.py:293]: Yeah I don't like how it splits the list into so many lines, but black is very opinionated. I didn't know about the excluding option --- I'll use that, thanks!

File: lhotse/bin/modes/recipes/dihard3.py
Comment (pzelasko) [inline lhotse/bin/modes/recipes/dihard3.py:19]:
```suggestion
    "-j", "--num-jobs",
```

**Review by pzelasko — APPROVED** [review]
I left just one comment; LGTM

**Comment by pzelasko** [comment]
Mm the CI is failing on linter - it seems the "Optional" symbol is not imported. Can you please re-run this before we merge?

**Comment by pzelasko** [comment]
Oh and one more thing -- can you add this to `lhotse/recipes/__init__.py` and list it in docs under available corpora `docs/corpora.rst`?

**Comment by desh2608** [comment]
@pzelasko I have made all the suggested changes.

**Comment by pzelasko** [comment]
Looks good to me. Should we merge?

**Comment by desh2608** [comment]
> Looks good to me. Should we merge?

@pzelasko yeah I think it can be merged once the checks complete.

**Pull request merged**
